File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/QueryByParsing.java
package io.codemodder.plugins.maven.operator;

import java.nio.file.Path;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.commons.lang3.builder.CompareToBuilder;
import org.apache.commons.lang3.text.StrSubstitutor;
import org.dom4j.Element;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Parses a Maven POM focusing on extracting dependency information without relying on any Maven
 * infrastructure.
 */
class QueryByParsing extends AbstractQueryCommand {

  private final Set<Dependency> dependencies = new LinkedHashSet<>();
  private final Set<Dependency> dependencyManagement =
      new TreeSet<>(
          new Comparator<Dependency>() {
            @Override
            public int compare(Dependency o1, Dependency o2) {
              if (o1 == o2) return 0;

              if (o1 == null) return 1;

              if (o2 == null) return -1;

              return new CompareToBuilder()
                  .append(o1.getGroupId(), o2.getGroupId())
                  .append(o1.getArtifactId(), o2.getArtifactId())
                  .toComparison();
            }
          });

  private final Map<String, String> properties = new LinkedHashMap<>();
  private final StrSubstitutor strSubstitutor = new StrSubstitutor(properties);

  @Override
  public void extractDependencyTree(Path outputPath, Path pomFilePath, ProjectModel c) {
    // Not implemented
  }

  /**
   * Executes the parsing strategy to extract dependency information from POM files.
   *
   * @param pm The ProjectModel containing input parameters for the operation.
   * @return `true` if the operation succeeds; otherwise, `false`.
   */
  @Override
  public boolean execute(ProjectModel pm) {
    /** Enlist all pom files given an hierarchy */
    List<POMDocument> pomFilesByHierarchy = pm.allPomFiles();
    Collections.reverse(pomFilesByHierarchy);

    for (POMDocument pomDocument : pomFilesByHierarchy) {
      updateProperties(pomDocument);
      updateDependencyManagement(pomDocument);
      updateDependencies(pomDocument);
    }

    this.result = dependencies;
    return true;
  }

  private void updateDependencyManagement(POMDocument pomDocument) {
    Collection<Dependency> dependencyManagementDependenciesToAdd = new ArrayList<>();

    Element dependencyManagementElement =
        pomDocument.getPomDocument().getRootElement().element("dependencyManagement");

    if (dependencyManagementElement != null) {
      List<Element> dependencyElements =
          dependencyManagementElement.element("dependencies").elements("dependency");

      for (Element dependencyElement : dependencyElements) {
        String groupId = getElementTextOrNull(dependencyElement, "groupId");
        String artifactId = getElementTextOrNull(dependencyElement, "artifactId");
        String version =
            Optional.ofNullable(dependencyElement.elementText("version")).orElse("UNKNOWN");
        String classifier = getElementTextOrNull(dependencyElement, "classifier");
        String packaging = getElementTextOrNull(dependencyElement, "packaging");

        try {
          version = strSubstitutor.replace(version);
        } catch (IllegalStateException e) {
          logger.warn("while interpolating version", e);
          version = "UNKNOWN";
        }

        Dependency dependency =
            new Dependency(groupId, artifactId, version, classifier, packaging, null);
        dependencyManagementDependenciesToAdd.add(dependency);
      }
    }

    this.dependencyManagement.addAll(dependencyManagementDependenciesToAdd);
  }

  private Dependency lookForDependencyManagement(String groupId, String artifactId) {
    for (Dependency dependency : dependencyManagement) {
      if (Objects.equals(dependency.getGroupId(), groupId)
          && Objects.equals(dependency.getArtifactId(), artifactId)) {
        return dependency;
      }
    }
    return null;
  }

  private void updateDependencies(POMDocument pomDocument) {
    Collection<Dependency> dependenciesToAdd = new ArrayList<>();

    Element dependenciesElement =
        pomDocument.getPomDocument().getRootElement().element("dependencies");

    if (dependenciesElement != null) {
      List<Element> dependencyElements = dependenciesElement.elements("dependency");

      for (Element dependencyElement : dependencyElements) {
        String groupId = getElementTextOrNull(dependencyElement, "groupId");
        String artifactId = getElementTextOrNull(dependencyElement, "artifactId");
        String version =
            Optional.ofNullable(dependencyElement.elementText("version")).orElse("UNKNOWN");

        Dependency proposedDependency = lookForDependencyManagement(groupId, artifactId);

        if (proposedDependency != null) {
          dependenciesToAdd.add(proposedDependency);
        } else {
          String classifier = getElementTextOrNull(dependencyElement, "classifier");
          String packaging = getElementTextOrNull(dependencyElement, "packaging");

          try {
            version = strSubstitutor.replace(version);
          } catch (IllegalStateException e) {
            logger.warn("while interpolating version", e);
            version = "UNKNOWN";
          }

          Dependency dependency =
              new Dependency(groupId, artifactId, version, classifier, packaging, null);
          dependenciesToAdd.add(dependency);
        }
      }
    }

    this.dependencies.addAll(dependenciesToAdd);
  }

  private String getElementTextOrNull(Element parent, String elementName) {
    Element child = parent.element(elementName);
    return child != null ? child.getText() : null;
  }

  /** Updates the Properties member variable based on whats on the POMDocument */
  private void updateProperties(POMDocument pomDocument) {
    Map<String, String> propsDefined = ProjectModel.propertiesDefinedOnPomDocument(pomDocument);

    for (Map.Entry<String, String> entry : propsDefined.entrySet()) {
      String key = entry.getKey();
      String value = entry.getValue();

      if (!value.matches(RE_INTERPOLATION.pattern())) {
        properties.put(key, value);
      }

      if (!value.matches(RE_INTERPOLATION.pattern())) {
        String newValue;
        try {
          Matcher matcher = RE_INTERPOLATION.matcher(value);
          StringBuffer resultBuffer = new StringBuffer();
          while (matcher.find()) {
            String variable = matcher.group();
            String replacement = strSubstitutor.replace(variable);
            matcher.appendReplacement(resultBuffer, Matcher.quoteReplacement(replacement));
          }
          matcher.appendTail(resultBuffer);
          newValue = resultBuffer.toString();
        } catch (IllegalStateException e) {
          LOGGER.warn("while replacing variables: ", e);
          newValue = value;
        }

        properties.put(key, newValue);
      }
    }
  }

  private static final Pattern RE_INTERPOLATION =
      Pattern.compile(".*\\$\\{[\\p{Alnum}.\\-_]+\\}.*");
  private static final Logger logger = LoggerFactory.getLogger(QueryByParsing.class);
}

File: pom-operator/src/main/java/io/github/pixee/maven/operator/QueryByParsing.kt
@file:Suppress("DEPRECATION")

package io.github.pixee.maven.operator

import org.apache.commons.lang3.builder.CompareToBuilder
import org.apache.commons.lang3.text.StrSubstitutor
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*

/**
 * this one is a bit more complex, as it intents to to a "best effort" attempt at parsing a pom
 * focusing only on dependency right now,  * without relying to any maven infrastructure at all
 */
class QueryByParsing : AbstractQueryCommand() {
    override fun extractDependencyTree(outputPath: File, pomFilePath: File, c: ProjectModel) {
        TODO("Not yet implemented")
    }

    private val dependencies: MutableSet<Dependency> = LinkedHashSet()

    private val dependencyManagement: MutableSet<Dependency> =
        TreeSet(object : Comparator<Dependency> {
            override fun compare(o1: Dependency?, o2: Dependency?): Int {
                if (o1 == o2)
                    return 0

                if (o1 == null)
                    return 1

                if (o2 == null)
                    return -1

                return CompareToBuilder().append(o1.groupId, o2.groupId)
                    .append(o1.artifactId, o2.artifactId).toComparison()
            }
        })

    private val properties: MutableMap<String, String> = LinkedHashMap()

    private val strSubstitutor = StrSubstitutor(properties)

    override fun execute(pm: ProjectModel): Boolean {
        /**
         * Enlist all pom files given an hierarchy
         */
        val pomFilesByHierarchy = pm.allPomFiles.reversed()

        for (pomDocument in pomFilesByHierarchy) {
            updateProperties(pomDocument)

            updateDependencyManagement(pomDocument)

            updateDependencies(pomDocument)
        }

        this.result = dependencies

        return true
    }

    private fun updateDependencyManagement(pomDocument: POMDocument) {
        val dependencyManagementDependenciesToAdd: Collection<Dependency> =
            pomDocument.pomDocument.//
            rootElement. //
            element("dependencyManagement")?. //
            element("dependencies")?. //
            elements("dependency")?. //
            map { dependencyElement ->
                val groupId = dependencyElement.element("groupId").text
                val artifactId = dependencyElement.element("artifactId").text

                val version = dependencyElement.element("version")?.text ?: "UNKNOWN"

                val classifier = dependencyElement.element("classifier")?.text
                val packaging = dependencyElement.element("packaging")?.text

                val versionInterpolated = try {
                    strSubstitutor.replace(version)
                } catch (e: java.lang.IllegalStateException) {
                    LOGGER.warn("while interpolating version", e)

                    "UNKNOWN"
                }

                Dependency(groupId, artifactId, versionInterpolated, classifier, packaging)
            }?.toList() ?: emptyList()

        this.dependencyManagement.addAll(dependencyManagementDependenciesToAdd)
    }

    fun lookForDependencyManagement(groupId: String, artifactId: String): Dependency? =
        this.dependencyManagement.firstOrNull { it.groupId == groupId && it.artifactId == artifactId }

    private fun updateDependencies(pomDocument: POMDocument) {
        val dependenciesToAdd: Collection<Dependency> =
            pomDocument.pomDocument.//
            rootElement. //
            element("dependencies")?. //
            elements("dependency")?. //
            map { dependencyElement ->
                val groupId = dependencyElement.element("groupId").text
                val artifactId = dependencyElement.element("artifactId").text

                val versionElement = dependencyElement.element("version")

                val proposedDependency = lookForDependencyManagement(groupId, artifactId)

                if (versionElement == null && null != proposedDependency) {
                    proposedDependency
                } else {
                    val version = versionElement?.text ?: "UNKNOWN"

                    val classifier = dependencyElement.element("classifier")?.text
                    val packaging = dependencyElement.element("packaging")?.text

                    val versionInterpolated = try {
                        strSubstitutor.replace(version)
                    } catch (e: java.lang.IllegalStateException) {
                        LOGGER.warn("while interpolating version", e)

                        "UNKNOWN"
                    }

                    Dependency(groupId, artifactId, versionInterpolated, classifier, packaging)
                }
            }?.toList() ?: emptyList()

        this.dependencies.addAll(
            dependenciesToAdd
        )
    }

    /**
     * Updates the Properties member variable based on whats on the POMDocument
     */
    private fun updateProperties(pomDocument: POMDocument) {
        val propsDefined = ProjectModel.propertiesDefinedOnPomDocument(pomDocument)

        propsDefined.entries.filterNot { it.value.matches(RE_INTERPOLATION) }
            .forEach {
                properties[it.key] = it.value
            }

        propsDefined.entries.filterNot { it.value.matches(RE_INTERPOLATION) }
            .forEach {
                val newValue = try {
                    strSubstitutor.replace(it.value)
                } catch (e: IllegalStateException) {
                    LOGGER.warn("while replacing variables: ", e)

                    it.value
                }

                properties.put(it.key, it.value)
            }
    }

    companion object {
        val RE_INTERPOLATION = Regex(""".*\$\{[\p{Alnum}.-_]+\}.*""")

        val logger: Logger = LoggerFactory.getLogger(QueryByParsing::class.java)
    }
}